### PR TITLE
Moderation post list 27

### PR DESCRIPTION
**Pull request by nationwide13**
This commit adds ability for posts to be created that are for moderators only. It's forum specific so each forum can have it's own private "mod chat"

It also creates an admin only forum so that the admins can have a site wide admin chat

### CODE REVIEW

File: templates/forum/forum.html.twig [hunk 1]
```diff
@@ -27,15 +27,17 @@
   {% spaceless %}
     <h1 class="page-heading forum-name-heading">
       <span class="forum-name-heading__prefix">/f/</span>
-      <strong class="forum-name-heading__name">{{ forum.name }}</strong>
+      <strong class="forum-name-heading__name">{{ forum.name }}{% if mod is defined %}/mod{% endif %}</strong>
     </h1>
   {% endspaceless %}
 
+  {% if mod is not defined %}
```
Comment (psineur) [inline templates/forum/forum.html.twig]: why do you want to remove sorting in mod-mode?

File: src/Repository/SubmissionRepository.php [hunk 2]
```diff
@@ -52,6 +54,31 @@ public function findFrontPageSubmissions(array $forums, string $sortBy, int $pag
     public function findForumSubmissions(Forum $forum, string $sortBy, int $page = 1) {
         $qb = $this->findSortedQb($sortBy)
             ->andWhere('s.forum = :forum')
+            ->andWhere('s.modThread = false')
+            ->setParameter('forum', $forum);
+
+        if ($sortBy === 'hot') {
+            PrependOrderBy::prepend($qb, 's.sticky', 'DESC');
+        }
```
Comment (psineur) [inline src/Repository/SubmissionRepository.php]: should this if-section be in separate PR/feature?

File: src/Repository/SubmissionRepository.php [hunk 3]
```diff
@@ -33,8 +33,10 @@ public function findFrontPageSubmissions(array $forums, string $sortBy, int $pag
 
         $qb = $this->findSortedQb($sortBy)
             ->where('IDENTITY(s.forum) IN (:forums)')
+            ->andWhere('s.modThread = false')
```
Comment (psineur) [inline src/Repository/SubmissionRepository.php]: this is a possibly security issue - we always have to add s.modThread = false everywhere. 

How about extracting a most common $qb creation into private function, maybe with modThread = false default param and adding all-caps warning - IF YOU FUCK THIS UP NORMAL USERS WILL SEE MOD POSTS

File: src/Repository/ForumRepository.php [hunk 4]
```diff
@@ -165,7 +176,7 @@ public function findForumNames($names) {
     public function findForumsInCategory(ForumCategory $category) {
         /* @noinspection SqlDialectInspection */
         $dql = 'SELECT f.id, f.name FROM '.Forum::class.' f '.
-            'WHERE f.category = :category '.
+            'WHERE f.category = :category and f.id > 0'.
```
Comment (psineur) [inline src/Repository/ForumRepository.php]: comment here or somewhere else why f.id > is necessary would be nice

File: src/Repository/ForumRepository.php [hunk 5]
```diff
@@ -86,11 +89,17 @@ public function findForumsByPage(int $page, string $sortBy) {
...
+            $dql =
+                'SELECT f.id, f.name FROM '.Forum::class.' f WHERE f IN ('.
+                    'SELECT IDENTITY(fs.forum) FROM '.ForumSubscription::class.' fs WHERE fs.user = ?1'.
+                ') ORDER BY f.normalizedName ASC';
+        } else {
+            $dql =
+                'SELECT f.id, f.name FROM '.Forum::class.' f WHERE f IN ('.
+                    'SELECT IDENTITY(fs.forum) FROM '.ForumSubscription::class.' fs WHERE fs.user = ?1 and f.id > 0'.
+                ') ORDER BY f.normalizedName ASC';
+        }
```
Comment (psineur) [inline src/Repository/ForumRepository.php]: let's refactor this into 

```
$nonAdminClauseIfAny = 'AND f.id > 0';
if ($user->isAdmin()) {
  $nonAdminClauseIfAny = '';
}

$dbql = BLABLA . $nonAdminClauseIfAny
```

this will prevent future edited-in-one-place-forgot-another bug and make it easier to distinguish admin/non-admin differences.

Actually you can probably extract #nonAdminClauseIfANy
to a function and use it everywhere sintead of 'f.id >0 ' string
Should minimize code-dupe

File: src/Repository/SubmissionRepository.php (hunk 2 above)
Comment (nationwide13) [reply src/Repository/SubmissionRepository.php]: This is just GIT being weird with diffs, this needs to be in here.

I duplicated this function but for the mod posts, and git got all weird

File: src/Repository/SubmissionRepository.php (hunk 3 above)
Comment (nationwide13) [reply src/Repository/SubmissionRepository.php]: Alright, I can pull that into a different function

File: templates/forum/forum.html.twig (hunk 1 above)
Comment (nationwide13) [reply templates/forum/forum.html.twig]: At first I couldn't convince the sortBy to work properly in this route. After more testing it looks like I've finally convinced it to work, so I'll update PR with that

File: src/Repository/ForumRepository.php (hunk 5 above)
Comment (nationwide13) [reply src/Repository/ForumRepository.php]: Meant to rewrite this before PR sorry

File: src/Repository/ForumRepository.php (hunk 4 above)
Comment (nationwide13) [reply src/Repository/ForumRepository.php]: Fixed in update

File: src/Repository/ForumRepository.php [hunk 6]
```diff
@@ -86,9 +89,13 @@ public function findForumsByPage(int $page, string $sortBy) {
...
     public function findSubscribedForumNames(User $user) {
         /* @noinspection SqlDialectInspection */
+        $showAdminForum = false;
+        if ($user->isAdmin()) {
+            $showAdminForum = true;
+        }
         $dql =
             'SELECT f.id, f.name FROM '.Forum::class.' f WHERE f IN ('.
-                'SELECT IDENTITY(fs.forum) FROM '.ForumSubscription::class.' fs WHERE fs.user = ?1'.
+                'SELECT IDENTITY(fs.forum) FROM '.ForumSubscription::class.' fs WHERE fs.user = ?1 andi' . ($showAdminForum) ? '' : ' f.id > 0' .
```
Comment (psineur) [inline src/Repository/ForumRepository.php]: andi?

File: src/Repository/SubmissionRepository.php [hunk 7]
```diff
@@ -102,6 +126,9 @@ public function findSortedQb($sortType) {
             throw new \InvalidArgumentException('Bad sort type');
         }
 
+        if (!$isAdmin) {
```
Comment (psineur) [inline src/Repository/SubmissionRepository.php]: add a note about dangers of changing this
